File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        answer = []
        subset = []

        def backtrack(index):
            if index >= len(nums):
                answer.append(subset.copy())
                return
            
            subset.append(nums[index])
            backtrack(index + 1)

            subset.pop()
            backtrack(index + 1)
            return
        
        backtrack(0)
        return answer

File: 0078-subsets/0078-subsets.java
class Solution {
    public void backtrack(int[] nums, int index, List<Integer> subset, List<List<Integer>> answer) {
        if (index >= nums.length) {
            answer.add(new ArrayList<>(subset));
            return;
        }
        subset.add(nums[index]);
        backtrack(nums, index + 1, subset, answer);

        subset.removeLast();
        backtrack(nums, index + 1, subset, answer);
        
        return;
    }

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> answer = new ArrayList<>();
        List<Integer> subset = new ArrayList<>();

        backtrack(nums, 0, subset, answer);

        return answer;
    }
}